### PR TITLE
Fix skill installation from URL

### PR DESCRIPTION
**Pull request by NeonDaniel**
Adds some bugfixes, logging, and unit tests to address issues encountered when trying to install skills from URLs.

## Changelog
Add authorname to SkillEntry generated in skill_entry_from_url
Add optional skill_dir param to install_skill_from_url
Add optional folder param to install_skill
Encapsulate and log errors in building SkillEntry.uuid
Add OSM test of install_skill_from_url
Add entry.uuid checks to TestSkillEntry
Add url to raised GithubInvalidUrl exceptions in utils.py

### CODE REVIEW

File: ovos_skills_manager/skill_entry.py [hunk 1]
```diff
@@ -28,9 +28,14 @@ def uuid(self):
...
         # github_repo.github_author , case insensitive
         # should be guaranteed to be unique
-        author = self.skill_author.lower()
-        repo = self.skill_folder.lower()
-        return repo + "." + author
+        try:
+            author = self.skill_author.lower()
+            repo = self.skill_folder.lower()
+            return repo + "." + author
+        except Exception as e:
```
Comment (ChanceNCounter) [inline ovos_skills_manager/skill_entry.py:35]: Devil or angel, you know I can't de-ci-ide...

I'm not opposed because I can't imagine the stuff in the `try` block triggering a system-level error. But I also know that nobody ever anticipates system-level errors.

I vote leave it, but keep "except Exception" in mind as a possible culprit in the event of bizarre tracebacks that end too soon.

File: ovos_skills_manager/skill_entry.py [hunk 2]
```diff
@@ -28,9 +28,14 @@ def uuid(self):
...
+            LOG.error(e)
+            LOG.error(f"author={self.skill_author} folder={self.skill_folder}")
```
Comment (ChanceNCounter) [inline ovos_skills_manager/skill_entry.py:37]: On the other hand, are these member variables guaranteed to exist? I don't see them being initialized anywhere.

File: ovos_skills_manager/skill_entry.py (hunk 1 above)
Comment (NeonDaniel) [reply ovos_skills_manager/skill_entry.py:35]: I considered making this a more specific type error, but I settled on this as being more concise.. Open to changing though

File: ovos_skills_manager/skill_entry.py (hunk 2 above)
Comment (NeonDaniel) [reply ovos_skills_manager/skill_entry.py:37]: They are properties, could raise the null exception in those property methods, but I wasn't sure if we always want to fail on null author or folder (wouldn't matter in search for example)

Comment (ChanceNCounter) [reply ovos_skills_manager/skill_entry.py:37]: I was thinking about the possibility of the `LOG` call raising an exception of its own:

```python3
>>> foo = SkillEntry({})
>>> print(f"{foo.skill_author}") # OK
None
>>> print(f"{foo.skill_folder}") # Here comes the pain
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/chance/hivemind-dev/ovos_skill_manager/ovos_skills_manager/skill_entry.py", line 114, in skill_folder
    return self.json.get("foldername") or self.url.split("/")[-1]
AttributeError: 'NoneType' object has no attribute 'split'
```

Incidentally, it also breaks `SkillEntry.__repr__`

File: ovos_skills_manager/skill_entry.py (hunk 1 above)
Comment (ChanceNCounter) [reply ovos_skills_manager/skill_entry.py:35]: Nah, I'm unrealistically confident in my assessment, ✅

**Comment by ChanceNCounter** [comment]
I'm just trusting in the thoroughness of the tests (see #39, apart from edge cases.) Other than that, however, I think the remaining unresolved box above is the only blocker. Otherwise, line by line, looks right, tests pass, and I've confirmed the custom exceptions now print the offending input, which is extremely helpful in the wild.

File: ovos_skills_manager/skill_entry.py (hunk 2 above)
Comment (NeonDaniel) [reply ovos_skills_manager/skill_entry.py]: Github auto-loaded all new comments except this one... I changed to raise an exception on skill_author not defined. If `repr` and `uuid` rely on non-null values I think it's fair to fail when either is null..

Comment (ChanceNCounter) [reply ovos_skills_manager/skill_entry.py]: It probably is, but it's also truthful to print an empty SkillEntry object.

Comment (ChanceNCounter) [reply ovos_skills_manager/skill_entry.py]: Oh, I see now. I definitely miscommunicated my concern.

I wasn't worried about the `skill_author` case succeeding, but rather about the `skill_folder` case failing. If these `LOG` calls raise exceptions, we're catching exceptions for the purpose of logging the occurrence and raising a specific error. If the values in question are null, I don't think the logger should raise an exception. I think it should log null and move onto the `SkillEntryError` at the end of the `except` block (not pictured above)

Comment (NeonDaniel) [reply ovos_skills_manager/skill_entry.py]: Oh, I see what you mean here; I've updated skill_folder to always return a string

Comment (ChanceNCounter) [reply ovos_skills_manager/skill_entry.py]: Now the author property is doing it ;)

Comment (NeonDaniel) [reply ovos_skills_manager/skill_entry.py]: whoops.. Fixed that one now. I'll add some SkillEntry tests...

**Review by ChanceNCounter — APPROVED** [review]

File: ovos_skills_manager/skill_entry.py [hunk 3]
```diff
@@ -242,8 +246,12 @@ def download(self, folder=None):
             ext = self.download_url.split(".")[-1]
         file = self.skill_folder + "." + ext
         url = self.download_url
+        skill_dirname = self.uuid
```
Comment (JarbasAl) [inline ovos_skills_manager/skill_entry.py:249]: nitpick, do we need to create this var? cant we use self.uuid in the couple lines below?

File: ovos_skills_manager/skill_entry.py [hunk 4]
```diff
@@ -317,11 +325,15 @@ def update(self, folder=None, default_branch="master", platform=None):
 
     def is_previously_installed(self, folder=None):
         folder = folder or get_skills_folder()
-        return isdir(join(folder, self.uuid))
+        skill_dirname = self.uuid
+        if not skill_dirname:
+            raise SkillEntryError(f"OSM installation of {self.skill_name or 'unknown skill'} failed!"
```
Comment (JarbasAl) [inline ovos_skills_manager/skill_entry.py:330]: does an exception make sense? if self.uuid does not exist then skill is not installed, and we just return False.

Installation failed doesnt really fit here since this will be mostly used before installing/updating

Comment (JarbasAl) [reply ovos_skills_manager/skill_entry.py:330]: `return isdir(join(folder, self.uuid)) if self.uuid else False`

Comment (ChanceNCounter) [reply ovos_skills_manager/skill_entry.py:330]: if self.uuid doesn't exist, doesn't that mean its entry is malformed?

Comment (JarbasAl) [reply ovos_skills_manager/skill_entry.py:330]: not necessarily, just means fields are missing from the data, unsure how that would ever happen in the wild however... Maybe someone creating an entry manually and missing the author field? If not built from url / local skill i can see this happening.

regardless is_installed is obviously False if thats missing, there is no need to throw an exception, anyone that tries to actually install gets the exception in it's proper place

We should revisit how to handle this use case later on, but feels like a different can of worms worth it's own PR

File: ovos_skills_manager/skill_entry.py (hunk 3 above)
Comment (ChanceNCounter) [reply ovos_skills_manager/skill_entry.py:249]: Some of its neighbors are also unneeded, but they predate this PR. I think simplifying these calls is worthwhile, but it doesn't *have* to be in this PR's scope. Could go either way.

For the record:

```python
        file = self.skill_folder + "." + ext
        url = self.download_url
        skill_dirname = self.uuid
        if not skill_dirname...
        return install_skill(url, folder, file, session=requests,
                             skill_folder_name=skill_dirname)
```

could all be reduced. On the other hand, the function might *benefit* from just moving `url` up, as that variable is read 3 times before this assignment.

Whole function could use a touch-up =P I wouldn't hold this PR over it. The broader error is breaking, and this needs to be released ASAP.

The other thing, fair enough.

File: ovos_skills_manager/skill_entry.py (hunk 4 above)
Comment (ChanceNCounter) [reply ovos_skills_manager/skill_entry.py:330]: I'm in the editor, so I'm just doing this one =P
